File: src/examples/inequality_scaling/inequality_scaling.cpp
// A problem that helps us determine how to scale inequality constrained
// optimization problems

#include "optizelle/optizelle.h"
#include "optizelle/vspaces.h"
#include "optizelle/json.h"
#include <iostream>
#include <iomanip>
#include <cstdlib>

// Grab Optizelle's natural type
typedef Optizelle::Natural Natural;

// Squares its input
template <typename Real>
Real sq(Real const & x){
    return x*x; 
}

// Define a simple objective where 
// 
// f(x) = 0.5 || x - c ||^2
//
struct MyObj : public Optizelle::ScalarValuedFunction <double,Optizelle::Rm> {
    typedef Optizelle::Rm <double> X;

    // Center of the quadratic objective
    X::Vector const & c;

    // Grab the center during constuction
    MyObj(X::Vector const & c_) : c(c_) {}

    // Evaluation 
    double eval(X::Vector const & x) const {
        auto acc = 0.;
        for(auto i=0;i<x.size();i++)
            acc+=sq(x[i]-c[i]);
        return 0.5*acc;
    }

    // Gradient
    void grad(
        X::Vector const & x,
        X::Vector & grad
    ) const {
        for(auto i=0;i<x.size();i++)
            grad[i]=x[i]-c[i];
    }

    // Hessian-vector product
    void hessvec(
        const X::Vector& x,
        const X::Vector& dx,
        X::Vector& H_dx
    ) const {
        for(Natural i=0;i<x.size();i++)
            H_dx[i]=dx[i];
    }
};

// Define simple inequalities 
//
// h(x) = x - lb 
//
struct MyIneq
    :public Optizelle::VectorValuedFunction<double,Optizelle::Rm,Optizelle::Rm>
{
    typedef Optizelle::Rm <double> X;
    typedef Optizelle::Rm <double> Z;

    // Lower bound 
    X::Vector const & lb;
    
    // Grab the lower bound during constuction
    MyIneq(X::Vector const & lb_) : lb(lb_) {}

    // z=h(x) 
    void eval(
        X::Vector const & x,
        Z::Vector & z
    ) const {
        for(auto i=0;i<x.size();i++)
            z[i]=x[i]-lb[i];
    }

    // z=h'(x)dx
    void p(
        X::Vector const & x,
        X::Vector const & dx,
        Z::Vector & z
    ) const {
        for(auto i=0;i<x.size();i++)
            z[i]=dx[i];
    }

    // xhat=h'(x)*dz
    void ps(
        X::Vector const & x,
        Z::Vector const & dz,
        X::Vector & xhat 
    ) const {
        for(auto i=0;i<x.size();i++)
            xhat[i]=dz[i];
    }

    // xhat=(h''(x)dx)*dz
    void pps(
        X::Vector const & x,
        X::Vector const & dx,
        Z::Vector const & dz,
        X::Vector & xhat 
    ) const {
        X::zero(xhat);
    }
};

int main(int argc,char* argv[]){
    // Read in the name for the input file
    if(argc!=2) {
        std::cerr << "inequality_scaling <parameters>" << std::endl;
        exit(EXIT_FAILURE);
    }
    auto fname = argv[1];

    // Create a type shortcut
    using Optizelle::Rm;

    // Generate an initial guess
    auto x = std::vector <double> (10);
    for(auto i=0;i<x.size();i++)
        x[i]=1. + pow(.1,i);

    // Allocate memory for the inequality multipler 
    auto z = std::vector <double> (10);

    // Create the center of the objective function
    auto c = std::vector <double> (10);
    for(auto i=0;i<c.size();i++)
        c[i]=-1.;

    // Create the lower bound for the problem
    auto lb = std::vector <double> (10);
    for(auto i=0;i<lb.size();i++)
        lb[i]=1.;

    // Create an optimization state
    Optizelle::InequalityConstrained <double,Rm,Rm>::State::t state(x,z);

    // Read the parameters from file
    Optizelle::json::InequalityConstrained <double,Optizelle::Rm,Optizelle::Rm>
        ::read(fname,state);
    
    // Create a bundle of functions
    Optizelle::InequalityConstrained <double,Rm,Rm>::Functions::t fns;
    fns.f.reset(new MyObj(c));
    fns.h.reset(new MyIneq(lb));

    // Solve the optimization problem
    Optizelle::InequalityConstrained <double,Rm,Rm>::Algorithms
        ::getMin(Optizelle::Messaging::stdout,fns,state);

    // Print out the reason for convergence
    std::cout << "The algorithm converged due to: " <<
        Optizelle::OptimizationStop::to_string(state.opt_stop) <<
        std::endl;

    // Print out the final answer
    std::cout << "The optimal point is: [" << std::endl;
    for(Natural i=0;i<state.x.size();i++)
        std::cout << std::scientific << std::setprecision(16)
            << state.x[i] << std::endl;
    std::cout << "]" << std::endl;

    // Write out the final answer to file
    Optizelle::json::InequalityConstrained<double,Rm,Rm>
        ::write_restart("solution.json",state);

    // Return that the program exited properly
    return EXIT_SUCCESS;
}

File: src/examples/inequality_scaling/inequality_scaling.py
# A problem that helps us determine how to scale inequality constrained
# optimization problems

import Optizelle 
import numpy
import sys

# Define a simple objective where 
# 
# f(x) = 0.5 || x - c ||^2
#
class MyObj(Optizelle.ScalarValuedFunction):
    # Grab the center of the objective
    def __init__(self, c):
        self.c = c 

    # Evaluation 
    def eval(self,x):
        diff = x - self.c
        return 0.5*numpy.dot(diff,diff)

    # Gradient
    def grad(self,x,grad):
        numpy.copyto(grad,x-self.c);

    # Hessian-vector product
    def hessvec(self,x,dx,H_dx):
        numpy.copyto(H_dx,dx);

# Define simple inequalities 
#
# h(x) = x - lb 
#
class MyIneq(Optizelle.VectorValuedFunction):
    # Grab the lower bound 
    def __init__(self, lb):
        self.lb = lb 

    # z=h(x) 
    def eval(self,x,z):
        numpy.copyto(z,x-lb);

    # z=h'(x)dx
    def p(self,x,dx,z):
        numpy.copyto(z,dx);

    # xhat=h'(x)*dz
    def ps(self,x,dz,xhat):
        numpy.copyto(xhat,dz);

    # xhat=(h''(x)dx)*dz
    def pps(self,x,dx,dz,xhat):
        xhat.fill(0.)

# Read in the name for the input file
if len(sys.argv)!=2:
    sys.exit("inequality_scaling.py <parameters>")
fname=sys.argv[1]

# Set the size
m = 10;

# Generate an initial guess 
x = numpy.array([1.+10**(-x) for x in xrange(1,m+1)])

# Allocate memory for the inequality multiplier 
z = numpy.array(m*[0.])

# Create the center of the objective function
c = numpy.array(m*[-1.])

# Create the lower bound for the problem
lb = numpy.array(m*[1.])

# Create an optimization state
state=Optizelle.InequalityConstrained.State.t(Optizelle.Rm,Optizelle.Rm,x,z)

# Read the parameters from file
Optizelle.json.InequalityConstrained.read(
    Optizelle.Rm,Optizelle.Rm,fname,state)

# Create a bundle of functions
fns=Optizelle.InequalityConstrained.Functions.t()
fns.f=MyObj(c)
fns.h=MyIneq(lb)

# Solve the optimization problem
Optizelle.InequalityConstrained.Algorithms.getMin(
    Optizelle.Rm,Optizelle.Rm,Optizelle.Messaging.stdout,fns,state)

# Print out the reason for convergence
print "The algorithm converged due to: %s" % (
    Optizelle.OptimizationStop.to_string(state.opt_stop))

# Print out the final answer
print "The optimal point is: ["
for i in xrange(m):
    print "%1.16e" % state.x[i]
print "]"

# Write out the final answer to file
Optizelle.json.InequalityConstrained.write_restart(
    Optizelle.Rm,Optizelle.Rm,"solution.json",state)
